scripts/migrate-state-exoscale-v4-v5.py

```diff
@@ -0,0 +1,455 @@
+
+
+def migrate_security_group_rules_lb(state: dict):
+    portspec_to_desc = {
+        "80-80": "Ingress controller HTTP",
+        "443-443": "Ingress controller HTTPS",
+        "6443-6443": "Kubernetes API",
+    }
+    sg_id = state["module.cluster.module.lb"]["exoscale_security_group"][
+        "load_balancers"
+    ]["instances"][0]["attributes"]["id"]
+    rules = state["module.cluster.module.lb"]["exoscale_security_group_rules"][
+        "load_balancers"
+    ]["instances"][0]["attributes"]["ingress"]
+
+    resid_base = "module.cluster.module.lb.exoscale_security_group_rule"
+    for r in rules:
+        for rid in r["ids"]:
+            rspec = parse_sg_rule_id(rid)
+            if rspec.portspec == "22623-22623":
+                # special case machine config server rule
+                resid = f"{resid_base}.load_balancers_machine_config_server[0]"
+            else:
+                desc = portspec_to_desc[rspec.portspec]
+                if rspec.source == "0.0.0.0/0":
+                    ipver = "4"
+                elif rspec.source == "::/0":
+                    ipver = "6"
+                else:
+                    print(
+                        "Unknown source {desc.source}, skipping rule", file=sys.stderr
+                    )
+                    continue
+                resid = f'{resid_base}.load_balancers_{rspec.proto}{ipver}["{desc}"]'
+
+            print(
+                f"Importing security group rule {resid} (Exoscale ID: {rspec.exo_id})",
+                file=sys.stderr,
+            )
+            terraform_import(resid, f"{sg_id}/{rspec.exo_id}")
+
+
+def main(state_json: Optional[str]):
+    default_node_groups = ["master", "infra", "storage", "worker"]
+
+    # Read Exoscale and additional worker group names from main.tf.json
+    zone = ""
+    main_tf_json = read_main_tf_json()
+    zone = main_tf_json["region"]
+    additional_worker_groups = list(
+        main_tf_json.get("additional_worker_groups", {}).keys()
+    )
+
+    if not zone:
+        print("Unable to extract Exoscale zone from main.tf.json", file=sys.stderr)
+        sys.exit(1)
+
+    if state_json:
+        with open(state_json, "r", encoding="utf-8") as f:
+            state = read_state_json(json.load(f))
+    else:
+        state = terraform_state_pull()
+
+    if "module.cluster" not in state:
+        print('Module "cluster" not found in state, exiting...', file=sys.stderr)
+        sys.exit(1)
+
+    if "module.cluster.module.lb" not in state:
+        print('Module "cluster.lb" not found in state, exiting...', file=sys.stderr)
+        sys.exit(1)
+
+    # Floating IPs
+    for eip in ["api", "ingress"]:
+        migrate_simple_resource(
+            state,
+            "module.cluster.module.lb",
+            "exoscale_ipaddress",
+            "exoscale_elastic_ip",
+            eip,
+            # pass index=-1 to omit [{index}] from the terraform resource identifier
+            index=-1,
```

Comment:
This is very magic to me 😅

Reply:
It's not particularly well engineered, but I didn't want to spend a lot of time to design a better interface which allows me to 
1) Migrate resources with index 0 (format `module.cluster.<type>.<name>[0]`) without having to explicitly specify it by setting the default value for the parameter to `0` -- this is the most common case
2) Migrate resources with index `i > 0` (format `module.cluster.<type>.<name>[i]`) by specifying the optional parameter
3) Migrate resources which don't have an index in their state specifier

Reply:
Added a docstring comment for `migrate_simple_resource()` which explains how it works